### DISCUSSION TITLE
❓ Why does Zed encourage fragmentation of Tree Sitter queries between editors?

### DISCUSSION THREAD

**Discussion by lougreenwood** [zed-industries/zed#16860] [Feature Requests]
I've been working on improving the support for [`Ember JS` in Zed](https://github.com/jylamont/zed-ember), this included updating the Tree Sitter queries.

Ember already has tree-sitter queries available [here](https://github.com/ember-tooling/tree-sitter-glimmer), these were initially put together to get support for highlight in nvim.

Since I started working on the Zed Ember improvements I realised that we're duplicating the TS queries between the Zed extension and the `tree-sitter-glimmer` package. I asked around in the Ember community and at least part of the reason is because editors may use different or custom capture groups.

Another thing I noticed was that in the `tree-sitter-glimmer` package - because it's targetting nvim - it uses `lua-match` instead of `match`.

Another thing that confuses me is that the Tree Sitter project itself says:

>[Overview](https://tree-sitter.github.io/tree-sitter/syntax-highlighting#overview)
**All of the files needed to highlight a given language are _normally included in the same git repository_ as the Tree-sitter grammar for that language** (for example, [tree-sitter-javascript](https://github.com/tree-sitter/tree-sitter-javascript), [tree-sitter-ruby](https://github.com/tree-sitter/tree-sitter-ruby)). In order to run syntax highlighting from the command-line, three types of files are needed

But Zed extensions encourage moving queries into the zed specific extension... so now a given language has 2 sets of TS queries to maintain.

Now... I could make the `tree-sitter-glimmer` a git submodule of the `zed-ember` extension and then alias the queries... but we still have the risk of incompatible capture groups used and also things like `lua-match` vs `match` (no idea if this is fine or not... `lua-match` seemed to work in Zed... but what other divergences could come in the future...?)

❓ So my question is... what's the deal here. Why does Zed encourage what seems like fragmentation and further maintenance burden. Why can't we just all use the same sets of queries?

**Comment by lougreenwood** [comment]
All of this is even more confusing when considering that Zed describes it's self as

> Code at the speed of thought – Zed is a high-performance, multiplayer code editor **from the creators of Atom and _Tree-sitter._**

**Comment by JosephTLyons** [comment]
Thanks for reaching out. We want to better support our community, and the best way for us to do that is through our official Discord. The team is active there and happy to help.

We’re closing the “Help and General Q&A” category on GitHub Discussions and directing all support questions to Discord going forward.

https://zed.dev/support
